### PR TITLE
fix(file source): line_agg preserves the context of the last line added

### LINKED ISSUE #14253: File source repeats input lines when multiline options are in use
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

When the `file` source is configured with `multiline` options, restarting Vector ends up regurgitating logs from the source. This has been verified with the configuration and sample input below. To reproduce, run once, stop vector after three seconds, and then run with the identical config again.

### Configuration

```text
[sources.api_logs]
type = "file"
glob_minimum_cooldown_ms = 2000
include = [ "$HOME/vector/*.log" ]
multiline.start_pattern = "^[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}"
multiline.condition_pattern = "^[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}"
multiline.mode = "halt_before"
multiline.timeout_ms = 1000

[sinks.vector_test]
type = "console"
inputs = [ "api_logs" ]
encoding.codec = "json"
```

### Version

v0.24.0

### Debug Output

```text
scripts/run-vector garden/config.toml
    Finished dev [unoptimized + debuginfo] target(s) in 0.32s
     Running `target/debug/vector --config garden/config.toml`
2022-09-01T22:08:40.934023Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info,buffers=info,kube=info"
2022-09-01T22:08:40.934270Z  INFO vector::app: Loading configs. paths=["garden/config.toml"]
2022-09-01T22:08:40.946870Z  INFO vector::topology::running: Running healthchecks.
2022-09-01T22:08:40.947120Z  INFO vector::topology::builder: Healthcheck: Passed.
2022-09-01T22:08:40.947378Z  INFO source{component_kind="source" component_id=api_logs component_type=file component_name=api_logs}: vector::sources::file: Starting file server. include=["/home/bruce.guenter/vector/garden/*.log"] exclude=[]
2022-09-01T22:08:40.947422Z  INFO vector: Vector has started. debug="true" version="0.25.0" arch="x86_64" build_id="none"
2022-09-01T22:08:40.954246Z  INFO source{component_kind="source" component_id=api_logs component_type=file component_name=api_logs}:file_server: file_source::checkpointer: Attempting to read legacy checkpoint files.
2022-09-01T22:08:40.963281Z  INFO source{component_kind="source" component_id=api_logs component_type=file component_name=api_logs}:file_server: vector::internal_events::file::source: Found new file to watch. file=/home/bruce.guenter/vector/garden/integrations.log
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"2022-08-28 12:56:44: sendData\nHTTP Method: POST\nRequestData: {xxx}\nResponse: {\n  \"response\": {\n    \"caseId\": \"12345\"\n  },\n  \"errors\": []\n}\n","source_type":"file","timestamp":"2022-09-01T22:08:40.965915524Z"}
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"2022-08-28 15:08:30: sendData\nHTTP Method: POST\nRequestData: {yyy}\nResponse: {\n  \"response\": {\n    \"caseId\": \"67890\"\n  },\n  \"errors\": []\n}\n","source_type":"file","timestamp":"2022-09-01T22:08:41.967441647Z"}
^C2022-09-01T22:08:51.092944Z  INFO vector: Vector has stopped.

> scripts/run-vector garden/config.toml
    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
     Running `target/debug/vector --config garden/config.toml`
2022-09-01T22:08:53.577732Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info,buffers=info,kube=info"
2022-09-01T22:08:53.578011Z  INFO vector::app: Loading configs. paths=["garden/config.toml"]
2022-09-01T22:08:53.591115Z  INFO vector::topology::running: Running healthchecks.
2022-09-01T22:08:53.591301Z  INFO vector::topology::builder: Healthcheck: Passed.
2022-09-01T22:08:53.591619Z  INFO vector: Vector has started. debug="true" version="0.25.0" arch="x86_64" build_id="none"
2022-09-01T22:08:53.591669Z  INFO source{component_kind="source" component_id=api_logs component_type=file component_name=api_logs}: vector::sources::file: Starting file server. include=["/home/bruce.guenter/vector/garden/*.log"] exclude=[]
2022-09-01T22:08:53.598707Z  INFO source{component_kind="source" component_id=api_logs component_type=file component_name=api_logs}:file_server: file_source::checkpointer: Loaded checkpoint data.
2022-09-01T22:08:53.601851Z  INFO source{component_kind="source" component_id=api_logs component_type=file component_name=api_logs}:file_server: vector::internal_events::file::source: Resuming to watch file. file=/home/bruce.guenter/vector/garden/integrations.log file_position=170
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"HTTP Method: POST","source_type":"file","timestamp":"2022-09-01T22:08:53.605442764Z"}
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"RequestData: {yyy}","source_type":"file","timestamp":"2022-09-01T22:08:53.605566318Z"}
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"Response: {","source_type":"file","timestamp":"2022-09-01T22:08:53.605667563Z"}
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"  \"response\": {","source_type":"file","timestamp":"2022-09-01T22:08:53.605790044Z"}
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"    \"caseId\": \"67890\"","source_type":"file","timestamp":"2022-09-01T22:08:53.605894423Z"}
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"  },","source_type":"file","timestamp":"2022-09-01T22:08:53.605985688Z"}
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"  \"errors\": []","source_type":"file","timestamp":"2022-09-01T22:08:53.606076962Z"}
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"}","source_type":"file","timestamp":"2022-09-01T22:08:53.606168519Z"}
{"file":"/home/bruce.guenter/vector/garden/integrations.log","host":"pf3","message":"","source_type":"file","timestamp":"2022-09-01T22:08:53.606270924Z"}
^C2022-09-01T22:08:55.343376Z  INFO vector: Vector has stopped.
2022-09-01T22:08:55.345159Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="api_logs, vector_test" time_remaining="59 seconds left"
```

### Example Data

```
2022-08-28 12:56:44: sendData
HTTP Method: POST
RequestData: {xxx}
Response: {
  "response": {
    "caseId": "12345"
  },
  "errors": []
}

2022-08-28 15:08:30: sendData
HTTP Method: POST
RequestData: {yyy}
Response: {
  "response": {
    "caseId": "67890"
  },
  "errors": []
}

```

### Additional Context

_No response_

### References

https://github.com/vectordotdev/vector/discussions/14158#discussioncomment-3515867

### PR DESCRIPTION
**Pull request by jches**
Updates the `line_agg` module to track the context of the last line added. By preserving the context as lines are added to the aggragator instead of only keeping the first context entry, the checkpointer in the file source will have access to the correct file offsets for merged lines when writing out checkpoints.

The tests here now use the source line index as the context object, and assert the index of the last merged line as the expected context of the merged lines. In the file source, context objects track the file offsets, and for checkpointing this will give the end_offset of the last merged line and fix #14253

### CODE REVIEW

**Review by bruceg — APPROVED** [review]
Thanks for the contribution, @jches. This looks good, and a great analysis of the root issue.

Do you have any sense if it would be difficult to write a unit test for the file source to check that this resolves the original issue there?

**Comment by jches** [comment]
At a glance, yeah it looks like there might be some existing unit tests in the file source that I could copy and adapt to test this

**Comment by bruceg** [comment]
Thanks, it would be great to have that test in there.

I think the soak performance regressions are related to a couple of issues currently fixed in `master`, so merging with that should resolve the issue.

**Comment by jches** [comment]
I added a test that _I think_ exercises this -- `cargo test` nearly crashes my laptop using all available memory though so I haven't been able to check it locally :sweat_smile:

**Comment by bruceg** [comment]
Running `cargo test` ends up running a bunch of link jobs in parallel, all of which take up lots of RAM, making it a nice DoS vector. 😜 You can avoid that by running:
```sh
cargo test --lib --no-default-features --features sources-file ::file
```

**Comment by jches** [comment]
Thanks for that tip! The test case needed a small change but should now cover this case :+1:

File: src/sources/file.rs [hunk 1]
```diff
@@ -1595,6 +1595,40 @@ mod tests {
         );
     }
 
+    #[tokio::test]
+    async fn test_multi_line_checkpointing() {
```
Comment (bruceg) [inline src/sources/file.rs:1599]: FWIW I confirmed that this fails on current master, so ✅

**Review by bruceg — APPROVED** [review]
Thanks again for this good work.

**Pull request merged**